File: upSolveSimulados/sim3/k.cpp
#include <bits/stdc++.h>

using namespace std;

#define fr(i,a,b) for(int (i) = (a); (i) < (b); ++(i))
#define rp(i,n) fr(i,0,n)
#define fi first
#define se second
#define all(v) (v).begin(), (v).end()
#define pb push_back
#define mt make_tuple
#define mp make_pair
#define sz(a) (int)(a.size())

typedef unsigned long long ull;
typedef long long ll;
typedef pair<int, int> ii;
typedef vector<int> vi;

const int inf = 0x3f3f3f3f;
const int neginf = 0xc0c0c0c0;
const ull limit = 1e18;
int r, s, m, d, n;


int main(){
	ios::sync_with_stdio(0);
	cin.tie(0);
	cin >> r >> s >> m >> d >> n;
	vector<int> brands(r);
	vector<vi> recipes(s+m+d);
	set<ii> naoRola;
	rp(i, r) cin >> brands[i];
	ull tot = 0;
	rp(i, s+m+d){
		int x, ing;
		cin >> x;
		rp(j, x){
			cin >> ing;
			recipes[i].pb(--ing);
		}
	}
	rp(i, n){
		int a, b;
		cin >> a >> b;
		if(a > b) swap(a,b);
		naoRola.insert(mp(--a,--b));
	}
	fr(i, 0, s){
		fr(j, s, s+m){
			fr(k, s+m, s+m+d){
				bool a = naoRola.find(mp(i,j)) != naoRola.end();
				bool b = naoRola.find(mp(i,k)) != naoRola.end();
				bool c = naoRola.find(mp(k,j)) != naoRola.end();
				if(a || b || c) continue;
				ull tmp = 1;
				ull prev = tmp;
				set<int> vaiRolar;
				vector<ull> multi;
				for(const int& v : recipes[i]) vaiRolar.insert(v);
				for(const int& v : recipes[j]) vaiRolar.insert(v);
				for(const int& v : recipes[k]) vaiRolar.insert(v);
				for(const int& v : vaiRolar) multi.pb((ull) brands[v]);	
				for(const ull& x : multi){
					prev = tmp;
					tmp *= x;
					if(tmp/x != prev){
						cout << "too many" << endl;
						return 0;
					}
				}
				if(limit - tot < tmp){
					cout << "too many" << endl;
					return 0;
				}
				tot += tmp;
			}
		}
	}
	if(tot <= limit)
		cout << tot << endl;
	else cout << "too many" << endl;
	return 0;
}

File: upSolveSimulados/sim3/k.py
from math import *

r, s, m, d, n = map(int, input().split())
brands = list(map(int, input().split()))
receitas = []
naoPode = dict()
for x in range(s+m+d):
	lista = list(map(int, input().split()))
	lista.pop(0)
	for k in range(len(lista)):
		lista[k] = lista[k] - 1
	receitas.append(lista)

print(receitas)

for x in range(n):
	a, b = map(int, input().split())
	if a > b:
		a,b = b,a
	naoPode[(a-1,b-1)] = 1


print(naoPode)
tot = 0
for i in range(s):
	for j in range(s, s+m):
		for k in range(s+m, s+m+d):
			if (i, j) in naoPode or (i, k) in naoPode or (k, j) in naoPode:
				continue
			tmp = 1
			temNoPrato = []
			for p in receitas[i]:
				temNoPrato.append(p)
			for p in receitas[j]:
				temNoPrato.append(p)
			for p in receitas[k]:
				temNoPrato.append(p)
			temNoPrato = set(temNoPrato)

			for item in temNoPrato:
				tmp = tmp * brands[item]
			tot += tmp
			
limit = 10 ** 18
print("{}".format(tot))
if tot > limit:
	print("too many")
else:
	print(tot)
